include/irr/asset/CShaderIntrospector.h

```diff
@@ -77,24 +71,17 @@ class CShaderIntrospector : public core::Uncopyable
 
     //In the future there's also going list of enabled extensions
     CShaderIntrospector(const IGLSLCompiler* _glslcomp, const SEntryPointStagePair& _ep) : m_glslCompiler(_glslcomp,core::dont_grab), m_entryPoint(_ep) {}
-    ~CShaderIntrospector() {
-        for (auto& introData : m_introspectionCache)
-            deinitIntrospectionData(introData.second);
-    }
 
-    const SIntrospectionData* introspect(const ICPUShader* _shader);
+    const CIntrospectionData* introspect(const ICPUShader* _shader);
 private:
-    SIntrospectionData doIntrospection(spirv_cross::Compiler& _comp, const SEntryPointStagePair& _ep) const;
-    void shaderMemBlockIntrospection(spirv_cross::Compiler& _comp, impl::SShaderMemoryBlock& _res, uint32_t _blockBaseTypeID, uint32_t _varID, const core::unordered_map<uint32_t, const SIntrospectionData::SSpecConstant*>& _mapId2sconst) const;
+    CIntrospectionData* doIntrospection(spirv_cross::Compiler& _comp, const SEntryPointStagePair& _ep) const;
+    void shaderMemBlockIntrospection(spirv_cross::Compiler& _comp, impl::SShaderMemoryBlock& _res, uint32_t _blockBaseTypeID, uint32_t _varID, const core::unordered_map<uint32_t, const CIntrospectionData::SSpecConstant*>& _mapId2sconst) const;
     size_t calcBytesizeforType(spirv_cross::Compiler& _comp, const spirv_cross::SPIRType& _type) const;
 
-    static void deinitIntrospectionData(SIntrospectionData& _data);
-    static void deinitShdrMemBlock(impl::SShaderMemoryBlock& _res);
-
 private:
     core::smart_refctd_ptr<const IGLSLCompiler> m_glslCompiler;
     SEntryPointStagePair m_entryPoint;
-    core::unordered_map<const ICPUShader*, SIntrospectionData> m_introspectionCache;
+    core::unordered_map<const ICPUShader*, core::smart_refctd_ptr<CIntrospectionData>> m_introspectionCache;
```

Comment:
keeping a cache with a key of raw-pointer will necessitate ICPUShaders letting know the `CShaderIntrospector` that they're being destroyed... maybe a smart pointer would be better (but then you need methods to kill the cache entries)

Reply:
well, pointer is just an integer and introspector isn't trying to access pointed address. There's no need of knowing whether the ICPUShader actually exist

Reply:
you do realize that if I create,drop and recreate ICPUShader I may end up with the same pointer but different shader contents ?

Reply:
hm, yea, so ill better make introspector hold them

Reply:
then you also need functions to purge this cache (so creating ICPUShader does not create actual memory leaks)

Reply:
@Crisspl status?

Reply:
done

Reply:
I can't see where in the code you're purging that cache, it just grows without bound.

Reply:
i meant key type is now smart_refctd_ptr

Reply:
great but need to be able to purge the cache totally as well as have the engine purge it periodically.

Reply:
>have the engine purge it periodically.

CShaderIntrospector isnt supposed to be one global object. It's not even refcounted. See how its used in COpenGLDriver::createGPUSpecializedShader()

Reply:
ok that clears it up for me

Reply:
resolved